### PR TITLE
58206: Replace usage of strpos with str_contains

### PR DESCRIPTION
**Pull request by Soean**
Replace usage of `strpos` with `str_contains`.

Trac ticket: https://core.trac.wordpress.org/ticket/58206

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-admin/includes/class-ftp.php [hunk 1]
```diff
@@ -797,7 +797,7 @@ function glob_pattern_match($pattern,$subject) {
 		$chunks=explode(';',$pattern);
 		foreach($chunks as $pattern) {
 			$escape=array('$','^','.','{','}','(',')','[',']','|');
-			while(strpos($pattern,'**')!==false)
+			while( str_contains( $pattern, '**' ) )
```
Comment (Soean) [inline src/wp-admin/includes/class-ftp.php:800]: Should we remove the spaces or should we stay with `strpos()` in this file?

Comment (Soean) [reply src/wp-admin/includes/class-ftp.php:800]:
```suggestion
			while(str_contains($pattern,'**') )
```

```suggestion
			while(strpos($pattern,'**')!==false)
```

**Comment by spacedmonkey** [comment]
@Soean Good start. I found some more places we could use this function. 

https://github.com/WordPress/wordpress-develop/blob/6d7430cfe1f8853b75787e03effe8f16f9242ad8/src/wp-admin/includes/theme.php#L1165

https://github.com/WordPress/wordpress-develop/blob/6d7430cfe1f8853b75787e03effe8f16f9242ad8/src/wp-admin/includes/theme.php#L1167

https://github.com/WordPress/wordpress-develop/blob/f3e65b1e441146d19f650971e4c932b358a7b10e/src/wp-admin/includes/ajax-actions.php#L3825

https://github.com/WordPress/wordpress-develop/blob/f3e65b1e441146d19f650971e4c932b358a7b10e/src/wp-admin/includes/file.php#L2334

https://github.com/WordPress/wordpress-develop/blob/6c47ed1dc697a81755d3892e41cede7b9a318d5b/src/wp-admin/includes/post.php#L1110

https://github.com/WordPress/wordpress-develop/blob/6c47ed1dc697a81755d3892e41cede7b9a318d5b/src/wp-admin/includes/post.php#L1119

https://github.com/WordPress/wordpress-develop/blob/f3e65b1e441146d19f650971e4c932b358a7b10e/src/wp-admin/includes/ajax-actions.php#L3362

https://github.com/WordPress/wordpress-develop/blob/f3e65b1e441146d19f650971e4c932b358a7b10e/src/wp-includes/class-wp-query.php#L4414

https://github.com/WordPress/wordpress-develop/blob/f3e65b1e441146d19f650971e4c932b358a7b10e/src/wp-includes/class-wp-query.php#L4414

https://github.com/WordPress/wordpress-develop/blob/f3e65b1e441146d19f650971e4c932b358a7b10e/src/wp-includes/class-wp-query.php#L4524

https://github.com/WordPress/wordpress-develop/blob/e4419d727ca99efff002ed7dbb70500e5093ed53/src/wp-includes/rss.php#L115

https://github.com/WordPress/wordpress-develop/blob/a028958dd6f01427fcf2fda5dd74c3a3a543b44a/src/wp-includes/blocks/navigation-submenu.php#L266

https://github.com/WordPress/wordpress-develop/blob/e4419d727ca99efff002ed7dbb70500e5093ed53/src/wp-includes/rss.php#L115

https://github.com/WordPress/wordpress-develop/blob/a028958dd6f01427fcf2fda5dd74c3a3a543b44a/src/wp-includes/blocks/navigation-submenu.php#L266

https://github.com/WordPress/wordpress-develop/blob/f3e65b1e441146d19f650971e4c932b358a7b10e/src/wp-includes/link-template.php#L4430

https://github.com/WordPress/wordpress-develop/blob/f3e65b1e441146d19f650971e4c932b358a7b10e/src/wp-includes/media.php#L1300

https://github.com/WordPress/wordpress-develop/blob/dcd1ba9330c23d3745c154eb11069ac6288c0665/src/wp-includes/option.php#L1263

https://github.com/WordPress/wordpress-develop/blob/f3e65b1e441146d19f650971e4c932b358a7b10e/src/wp-includes/pluggable.php#L1236

https://github.com/WordPress/wordpress-develop/blob/f3e65b1e441146d19f650971e4c932b358a7b10e/src/wp-includes/post.php#L6106

https://github.com/WordPress/wordpress-develop/blob/e4419d727ca99efff002ed7dbb70500e5093ed53/src/wp-includes/rss.php#L115

https://github.com/WordPress/wordpress-develop/blob/6c47ed1dc697a81755d3892e41cede7b9a318d5b/src/wp-includes/user.php#L3061

https://github.com/WordPress/wordpress-develop/blob/f3e65b1e441146d19f650971e4c932b358a7b10e/src/wp-includes/rest-api/class-wp-rest-request.php#L309

**Comment by Soean** [comment]
@spacedmonkey Thanks, I added the post.php line.

In the first step I only changed the `strpos( ) === false` and `strpos( ) !== false`, as described in the ticket . At the others we have to look at very closely, because if a string starts with a substring, `strpos` returns `0` which is `false` and `str_contains` returns `true`.

```php
$string = '/wp-content'
if ( strpos( $string, '/' ) ) { // 0 -> false
if ( str_contains( $string, '/' )  ) { // -> true
```

**Review by costdev — APPROVED** [review]
Thanks for the PR @Soean! I've reviewed all instances and they correctly change `strpos() === false` to `! str_contains()`, and `strpos() !== false` to `str_contains()`.

Regarding loose `strpos()` checks, I agree that this needs consideration and a consistent approach to determining the appropriate change(s). To make use of modern functions, both `str_contains()` and `str_starts_with()` may be necessary in certain circumstances to protect backward compatibility and prevent possible security regressions.

Using both of these functions may negate the performance benefit (that needs to be checked), but may still provide a benefit to readability.

For the other cases with a clean change from  `strpos() === false` to `! str_contains()`, and `strpos() !== false` to `str_contains()` though, I think this PR is a quick win that's ready for commit consideration.

**Review by spacedmonkey — APPROVED** [review]

File: src/wp-admin/includes/class-ftp.php (hunk 1 above)
Comment (peterwilsoncc) [reply src/wp-admin/includes/class-ftp.php:800]: It looks like the library is adopted to I think it's fine to keep the change here. The spaces can remain,

File: src/wp-includes/functions.php [hunk 2]
```diff
@@ -5962,13 +5962,13 @@ function apache_mod_loaded( $mod, $default_value = false ) {
 
 	if ( empty( $loaded_mods )
 		&& function_exists( 'phpinfo' )
-		&& false === strpos( ini_get( 'disable_functions' ), 'phpinfo' )
+		 && ! str_contains( ini_get( 'disable_functions' ), 'phpinfo' )
```
Comment (peterwilsoncc) [inline src/wp-includes/functions.php]:
```suggestion
		&& ! str_contains( ini_get( 'disable_functions' ), 'phpinfo' )
```

File: src/wp-includes/functions.php [hunk 3]
```diff
@@ -7098,9 +7098,9 @@ function _device_can_upload() {
 
 	$ua = $_SERVER['HTTP_USER_AGENT'];
 
-	if ( strpos( $ua, 'iPhone' ) !== false
-		|| strpos( $ua, 'iPad' ) !== false
-		|| strpos( $ua, 'iPod' ) !== false ) {
+	if ( str_contains( $ua, 'iPhone' )
+		 || str_contains( $ua, 'iPad' )
+		 || str_contains( $ua, 'iPod' ) ) {
```
Comment (peterwilsoncc) [inline src/wp-includes/functions.php]:
```suggestion
		|| str_contains( $ua, 'iPad' )
		|| str_contains( $ua, 'iPod' ) ) {
```

**Review by peterwilsoncc — COMMENTED** [review]
A few style nits where soem whitespace has leaked in but this looks good.

Code review notes:
- checked conditions match 
- checked no external libraries were updated

File: src/wp-includes/functions.php (hunk 2 above)
Comment (Soean) [reply src/wp-includes/functions.php]: Good catch, thanks!

**Comment by spacedmonkey** [comment]
@Soean Do you mind rebase from trunk. they are some new str_pos that have been added in newer commits.

**Comment by SergeyBiryukov** [comment]
Thanks for the PR! Merged in [r55988](https://core.trac.wordpress.org/changeset/55988).

**Pull request closed without merging**